src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,262 @@
+using System.Threading.Tasks;
+using Amazon;
+using Amazon.Runtime;
+using Amazon.S3;
+using Amazon.S3.Model;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress.S3
+{
+    /// <summary>
+    /// Egress provider for egressing stream data to the s3 storage.
+    /// </summary>
+    internal class S3StorageEgressProvider : EgressProvider<S3StorageEgressProviderOptions>
+    {
+        public S3StorageEgressProvider(ILogger<S3StorageEgressProvider> logger) : base(logger)
+        {
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<CancellationToken, Task<Stream>> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await using var stream = await action(token);
+                var resourceId = await UploadToStorageAsync(options, stream, artifactSettings, token);
+                Logger?.EgressProviderSavedStream(EgressProviderTypes.S3Storage, resourceId);
+                return resourceId;
+            }
+            catch (EgressException)
+            {
+                throw;
+            }
+            catch (Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<Stream, CancellationToken, Task> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                var client = await CreateClientAsync(options, token);
+                var uploadId = await InitMultiPartUploadAsync(client, options.BucketName, artifactSettings, token);
+                await using var stream = new MultiPartUploadStream(client, options.BucketName, artifactSettings.Name, uploadId, options.CopyBufferSize!.Value);
+                
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await action(stream, token);
+                await stream.FinalizeAsync(token); // force to push the last part
+                stream.Close();
+
+                if (stream.Parts.Count == 0)
+                {
+                    Logger?.LogDebug(Strings.Message_EgressS3NoContent);
+                    return string.Empty; // nothing to return here
+                }
+
+                await CompleteMultiPartUploadAsync(client, options.BucketName, uploadId, artifactSettings, stream.Parts, token);
+
+                var resourceId = client.GetPreSignedURL(new GetPreSignedUrlRequest { BucketName = options.BucketName, Key = artifactSettings.Name, Expires = DateTime.UtcNow.AddHours(1) });
```

Comment:
So what should I return here otherwise?

Reply:
We would prefer to have this return a URL or something path-like in order to be informative of where the blob exists. We've done this for the Azure Blob egress by removing the query parameter portion of the URL since the host name and path is the part that we actually want to return: https://github.com/dotnet/dotnet-monitor/blob/main/src/Tools/dotnet-monitor/Egress/AzureBlob/AzureBlobEgressProvider.cs#L323

Reply:
We also wanted to return this as an URI as we need to wanted to be similar to other egress providers which also return a concrete "path" to the upload resource. I would also prefer to keep it as it is.

Reply:
Nevertheless I added the suggested settings to let the user decide what is preferred